### PR TITLE
SINGA-254 Implement Adam for V1

### PR DESCRIPTION
**Pull request by nudles**
Implemented Adam for pysinga.
Added an argument to control the applying of regularization and constraint
for parameters, which is necessary for BN layers as the mean and std
should not be updated by updaters. The effect is not obvious for other
udpaters, but is very significant for Adam.

Tested Adam for alexnet on cifar10, and the accuracy was 0.8 (SGD is
0.82); It was also not as good as SGD on VGG net (0.92); May need to
tune the learning rate.

To test this PR, 
1. please replace the optimizer from SGD to Adam in examples/cifar10/train.py
2. pass a constant learning rate (e.g 0.01) to apply_with_lr, instead of the learning rate generation function
3. add the iteration ID as the last argument of apply_with_lr.

### CODE REVIEW

File: python/singa/optimizer.py [hunk 1]
```diff
@@ -251,47 +287,70 @@ class AdaGrad(Optimizer):
...
+        if name not in self.m or name not in self.v:
+            self.m[name] = tensor.Tensor(grad.shape, grad.device, grad.dtype)
+            self.m[name].set_value(0)
+            self.v[name] = tensor.Tensor(grad.shape, grad.device, grad.dtype)
+            self.v[name].set_value(0)
+
+        self.m[name] *= self.beta_1
+        tensor.axpy(1 - self.beta_1, grad, self.m[name])
+        self.v[name] *= self.beta_2
+        tensor.axpy(1 - self.beta_1, tensor.square(grad), self.v[name])
```
Comment (ijingo) [inline python/singa/optimizer.py]: typo: should be "tensor.axpy(1 - self.beta_2, tensor.square(grad), self.v[name])"
